src/core/src/op/non_zero.cpp

```diff
@@ -2,179 +2,147 @@
-        rc = evaluate_nonzero_execute<INPUT_ET, element::Type_t::a>(__VA_ARGS__); \
-    } break
-
-template <element::Type_t INPUT_ET>
-bool evaluate(const HostTensorPtr& input, const HostTensorPtr& output) {
-    bool rc = true;
-    switch (output->get_element_type()) {
-        TYPE_OUT_CASE(i64, input, output);
-        TYPE_OUT_CASE(i32, input, output);
-    default:
-        rc = false;
-        break;
-    }
-
-    return rc;
-}
-#undef TYPE_OUT_CASE
-bool evaluate_nonzero(const HostTensorPtr& input, const HostTensorPtr& output) {
-    bool rc = true;
-
-    switch (input->get_element_type()) {
-        OPENVINO_TYPE_CASE(evaluate_nonzero, boolean, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, i8, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, i16, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, i32, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, i64, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, u8, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, u16, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, u32, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, u64, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, bf16, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, f16, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, f32, input, output);
-        OPENVINO_TYPE_CASE(evaluate_nonzero, f64, input, output);
-    default:
-        rc = false;
-        break;
-    }
-    return rc;
+    return std::make_shared<v3::NonZero>(new_args.at(0), m_output_type);
 }
-}  // namespace
-}  // namespace nonzero
 
-bool op::v3::NonZero::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const {
+bool NonZero::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
     OV_OP_SCOPE(v3_NonZero_evaluate);
-    return nonzero::evaluate_nonzero(inputs[0], outputs[0]);
+
+    const auto& input = inputs[0];
+    auto& output = outputs[0];
+    using namespace ov::element;
+    const auto& input_shape = input.get_shape();
+    return IfTypeOf<boolean, bf16, f16, f32, f64, i8, i16, i32, i64, u8, u16, u32, u64>::apply<non_zero::Evaluate>(
+        input.get_element_type(),
+        input,
+        input_shape,
+        input_shape.size(),
+        output);
 }
 
-bool op::v3::NonZero::has_evaluate() const {
+bool NonZero::has_evaluate() const {
     OV_OP_SCOPE(v3_NonZero_has_evaluate);
     switch (get_input_element_type(0)) {
-    case ngraph::element::i8:
-    case ngraph::element::i16:
-    case ngraph::element::i32:
-    case ngraph::element::i64:
-    case ngraph::element::u8:
-    case ngraph::element::u16:
-    case ngraph::element::u32:
-    case ngraph::element::u64:
-    case ngraph::element::bf16:
-    case ngraph::element::f16:
-    case ngraph::element::f32:
-    case ngraph::element::f64:
+    case element::boolean:
```

Comment:
Looks like `boolean` was not in original list of supported types. Why it was added? How it will work with

Reply:
`boolean` was (and still is) handled in `evaluate` method. There's even a test for that. I assumed that these two method should be aligned.